[observability] Add alert when workspaces don't start

princerachit opened the pull request:
## Description
<!-- Describe your changes in detail -->

For the given formula the alerts can fire very often for xl clusters. To disable it I have created a custom pagerduty eventrule to suppress alerts being fired from the xl clusters. Refer to this [internal thread](https://gitpod.slack.com/archives/C0201TXMV54/p1650443926088999)

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/8448

## How to test
<!-- Provide steps to test this PR -->
The alert generation itself does not need to be tested as this is a standard addition to alerts. However , you can review this as part of review:

### If alerts are suppressed for xl clusters
**NOTE**: Please make sure you inform the oncall about your tests, or override yourself to be the oncall before running the test

* Start a workspace and fetch kubeconfig of a cluster from which you want to generate the alert.
* port forward alertmanager service in the terminal 

```
kubectl port-forward svc/alertmanager-main 9093:9093
```

* From a different terminal run these one by one and check pager duty.


1. Create an alert with right name but cluster is not xl -> You should see an incident in pagerduty

```console
curl -d '[{"labels": {"Alertname": "GitpodWorkspacesNotStarting", "severity": "critical", "cluster": "us41", "container": "ws-daemon", "instance": "10.20.169.66:8443"}}]' http://localhost:9093/api/v1/alerts
```

2. Create an alert with right name and cluster set to xl -> You should NOT see an incident in pagerduty

```console
curl -d '[{"labels": {"Alertname": "GitpodWorkspacesNotStarting", "severity": "critical", "cluster": "us41xl", "container": "ws-daemon", "instance": "10.20.169.66:8443"}}]' http://localhost:9093/api/v1/alerts
```

3. Create an alert with wrong name but cluster is xl -> You should see an incident in pagerduty

```console
curl -d '[{"labels": {"Alertname": "G1itpodWorkspacesNotStarting", "severity": "critical", "cluster": "us41xl", "container": "ws-daemon", "instance": "10.20.169.66:8443"}}]' http://localhost:9093/api/v1/alerts
```


### Additional References
You can use this formula to plot graph (you would need to set cluster variable in dashboard settings, refer to the success criteria dashboard for the formula)

```
count(avg_over_time(gitpod_workspace_regular_not_active_percentage{cluster=~"$cluster"}[10m])) * count(rate(gitpod_ws_manager_workspace_startup_seconds_sum{type="REGULAR",cluster=~"$cluster"}[10m]) == 0)
```

<details>
    <summary>You can alternatively just import this dashboard in Grafana.</summary>
https://gist.github.com/princerachit/c28960e5afb4a316e0cfc35d3d674987
</details>

#### Past Data
Here is a list of various cluster screenshots. Each point in the cluster indicates that this alert would have been fired

<details>
    <summary>us38</summary>
<img width="1256" alt="image" src="https://user-images.githubusercontent.com/32481722/165066417-ca6df9ea-c986-4f2f-bd26-0af5ca90ab3f.png">
</details>


<details>
    <summary>eu38</summary>

<img width="1256" alt="image" src="https://user-images.githubusercontent.com/32481722/165066548-bd85ee54-d539-486f-b47e-c0e3bb56ed5c.png">
</details>


<details>
    <summary>us40</summary>
<img width="1256" alt="image" src="https://user-images.githubusercontent.com/32481722/165066696-36d0a87f-313e-416f-88f9-f38c5fd48b4a.png">
</details>

<details>
    <summary>eu40</summary>
<img width="1256" alt="image" src="https://user-images.githubusercontent.com/32481722/165066770-b4d501c2-b022-4bc0-b367-e38524f30fbe.png">
</details>

<details>
    <summary>us39</summary>

<img width="1256" alt="image" src="https://user-images.githubusercontent.com/32481722/165066868-ec7b1b6b-84fb-4f53-b6bb-05f28d78e487.png">
</details>

<details>
    <summary>eu39</summary>

<img width="1256" alt="image" src="https://user-images.githubusercontent.com/32481722/165066996-ae165064-affb-4650-9399-5d4c3b3e5ac7.png">
</details>

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
None
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

kylos101 commented on operations/observability/mixins/workspace/rules/components/workspaces/alerts.libsonnet:
```diff
@@ -73,6 +73,23 @@
+            labels: {
+              severity: 'critical',
+            },
+            'for': '10m',
```
Hi @princerachit does it make sense to leave this at 10m, when the expression is inspecting rate for 5m? :thinking: Seems odd to me, but, maybe I am missing something.

princerachit replied:
I can change this but it does not make any difference, afterall we are checking if the value is zero

kylos101 replied:
Can you share if this would have triggered in the last 30 or 60 days with support detail? 

My understanding is `for` controls how long an expression must be true before the alert fires. And the underlying expression is looking for a window of time where the sum of starts is 0.

I would expect this alert to fire after 20 minutes. For 10 minutes, the following must be true:
1. regular not active has been > than zero
2. no workspaces have started in the last 10 minutes

I think it would have triggered if we were using 5m and 5m, triggered [here](https://grafana.gitpod.io/?orgId=1&var-datasource=VictoriaMetrics&var-cluster=us38xl2&from=1649419599314&to=1649422342364).

![image](https://user-images.githubusercontent.com/1272076/162476702-7fccc8ed-739e-43a9-a59f-852a3dfe4736.png)

kylos101 replied:
@princerachit marked as unresolved so I can get your [feedback](https://github.com/gitpod-io/gitpod/pull/9190#discussion_r846264519).

ArthurSens replied:
> I would expect this alert to fire after 20 minutes.

That is 100% correct, 

When using `rate(metric[10m])`, the value will be calculated using the latest 10 minutes of data. This means that the value cannot increase for the last 10 minutes for this value ever becomes 0. Adding a `for` to the alert will just increase the time for the alert to be triggered.

For the alert to be triggered in 10 minutes you can:
* rate[5m] + for: 5m
* rate[10m]

ArthurSens replied:
[Some extra info on rates](https://promlabs.com/blog/2021/01/29/how-exactly-does-promql-calculate-rates) :)

ArthurSens commented on operations/observability/mixins/workspace/rules/components/workspaces/alerts.libsonnet:
```diff
@@ -73,6 +73,23 @@
+            expr: |||
+              gitpod_workspace_regular_not_active_percentage > 0.0
+              AND
+              sum(rate(gitpod_ws_manager_workspace_startup_seconds_bucket{type="REGULAR",le="+Inf"}[10m])) == 0
```
```suggestion
              sum(rate(gitpod_ws_manager_workspace_startup_seconds_count{type="REGULAR"}[10m])) == 0
```

Every prometheus histogram comes with a counter 🙂

kylos101 commented:
@ArthurSens :pray: thank you for taking a gander!

princerachit commented:
I have fixed the calculation issues and made it similar to the way we have been doing e.g. 1m for 10m.

I have also updated metrics so that formula becomes simple and intutive.

You can refer to the graph url that I just added in the description @kylos101 to see how often this alert can get triggered

kylos101 commented:
@princerachit we must make it large enough so that it doesn't trigger for XL clusters. Is 20 minutes enough time? Please look in Grafana and update accordingly.

kylos101 commented:
> @princerachit we must make it large enough so that it doesn't trigger for XL clusters. Is 20 minutes enough time? Please look in Grafana and update accordingly the alert accordingly, sharing a picture so the reviewer has context as to why you set the value.

Hey @princerachit , sounds like the XL clusters have gaps that are too large. Can we exclude them from this alert for now? That way you can ship a :skateboard:

princerachit commented:
I have reached out to the team internally on how to implement this filtering of xl clusters. Based on the promethus configuration that we have, it seems impossible to disable alert from prometheus running in the cluster itself. This is because we use external_labels to add cluster information.

kylos101 commented:
@princerachit can you share a short loom demonstrating the test? I'm not able to recreate.

princerachit commented:
@kylos101 I have added all the information in the description. I have also added screenshots.
I have not captured this in loom as the critical channel was already polluted with my tests.

kylos101 commented:
Thank you, @princerachit . This is an excellent start, and we can tune from here. 

Can you update the runbook, to say something like, please try starting workspaces in the context cluster? If the on-caller can start workspaces in the context cluster (or not) I think that will influence whether or not they escalate further. It doesn't mention to do this, and it seems like a pretty good first step. For example, they might have to use a VPN to access the US or EU cluster.

kylos101 reviewed (approved):
:+1:
